esp/services/ws_dfu/ws_dfuService.cpp

```diff
@@ -5885,4 +5897,207 @@ int CWsDfuEx::GetIndexData(IEspContext &context, bool bSchemaOnly, const char* i
+void getFileMeta(IPropertyTree &metaInfo, IDistributedFile &file, IUserDescriptor *user, const char *keyPairName, IConstDFUReadAccessRequest &req)
+{
+    metaInfo.setProp("version", securityInfoVersion);
+    metaInfo.setProp("logicalFilename", file.queryLogicalName());
+    metaInfo.setProp("jobId", req.getJobId());
+    metaInfo.setProp("accessType", req.getAccessTypeAsString());
+    StringBuffer userStr;
+    if (user)
+        metaInfo.setProp("user", user->getUserName(userStr).str());
+
+    const char *clusterName = req.getCluster(); // can be null
+    Owned<IFileDescriptor> fDesc = file.getFileDescriptor(clusterName);
+
+#ifdef _USE_OPENSSL
+    if (keyPairName) // without it, meta data is not encrypted
+    {
+        metaInfo.setProp("keyPairName", keyPairName);
+
+        unsigned expirySecs = req.getExpirySeconds();
+
+        // JCSMORE - make this configurable
+        if (expirySecs > 86400)
+            expirySecs = 86400; //24 hours
+
+        // setup "expiryTime"
+        time_t simple;
+        time(&simple);
+        simple += expirySecs;
+        CDateTime expiryDt;
+        expiryDt.set(simple);
+        StringBuffer expiryTimeStr;
+        expiryDt.getString(expiryTimeStr);
+        Owned<IPropertyTree> secureMetaInfo = createPTreeFromIPT(&metaInfo);
+        secureMetaInfo->setProp("expiryTime", expiryTimeStr);
+
+        extractFilePartInfo(*secureMetaInfo, *fDesc);
+
+        // create random AES key and IV
+        char randomAesKey[aesKeySize];
+        char randomIV[aesBlockSize];
+        fillRandomData(aesKeySize, randomAesKey);
+        fillRandomData(aesBlockSize, randomIV);
+
+        // 1st serialize to JSON
+        StringBuffer secureMetaInfoJson;
+        toJSON(secureMetaInfo, secureMetaInfoJson);
+
+        // 2nd compress
+        MemoryBuffer compressedSecureMetaInfoMb;
+        fastLZCompressToBuffer(compressedSecureMetaInfoMb, secureMetaInfoJson.length(), secureMetaInfoJson.str());
+
+        // 3rd encrypt with AES key
+        MemoryBuffer encryptedSecureMetaInfoMb;
+        aesKeyEncrypt(encryptedSecureMetaInfoMb, compressedSecureMetaInfoMb.length(), compressedSecureMetaInfoMb.bytes(), randomAesKey, randomIV);
+
+        metaInfo.setPropBin("secureInfo", encryptedSecureMetaInfoMb.length(), encryptedSecureMetaInfoMb.bytes());
+
+        // 4th encrypt AES key with public *DAFILESRV* key
+        VStringBuffer publicKeyFName("%s.pub.pem", keyPairName); // JCSMORE - could cache loaded keys
+        Owned<CLoadedKey> publicKey = loadPublicKeyFromFile(publicKeyFName, nullptr);
+        MemoryBuffer encryptedAesKeyMb;
+        publicKeyEncrypt(encryptedAesKeyMb, aesKeySize, randomAesKey, *publicKey);
+
+        /* perhaps should combine into single prob..
+         * Does IV need to be generated/sent each time?
+         */
+        metaInfo.setPropBin("securityKey", encryptedAesKeyMb.length(), encryptedAesKeyMb.bytes());
+        metaInfo.setPropBin("securityIV", aesBlockSize, randomIV);
```

Comment:
@afishbeck - please look over this. This is the standard approach to using public/private key to secure a oneshot secret symmetric [aes] key to encrypt the data, such that the receiver can use it's private key to get the symmetric key to decrypt.

This does:
1) generating a random 256bit aes key and random IV (block size - 128bit).
2) encrypting the secure meta info (which is compressed JSON blob) with the AES key from (1).
3) encrypting the AES key with RSA and adding it to the envelope for the client to use.

I have the other side implemented (dafilesrv) - will be in a separate PR shortly.
Other side which decrypts the AES key with it's private RSA key and can then securely decode the meta info.

As I think I comment somewhere, nothing new, but as a reminder, this implies that the authority server (esp) has the public key of the dafilesrv(s) and as it stands, because it is one blob destined for many dafilesrv's, all the dafilesrv relating to a file (group) must use the same key.
To break that restriction, and have every dafilesrv use it's own key (and for esp to have all their public keys), would require a different scheme, where blobs destined for each dafilesrv and encoded separately etc.